File: java/yass/main/ch/softappeal/yass/transport/PacketSerializer.java
package ch.softappeal.yass.transport;

import ch.softappeal.yass.remote.Message;
import ch.softappeal.yass.remote.session.Packet;
import ch.softappeal.yass.serialize.CompositeSerializer;
import ch.softappeal.yass.serialize.Reader;
import ch.softappeal.yass.serialize.Serializer;
import ch.softappeal.yass.serialize.Writer;

/**
 * Default {@link Serializer} for {@link Packet}.
 */
public final class PacketSerializer extends CompositeSerializer {

    /**
     * @param messageSerializer A {@link Serializer} for {@link Message}.
     */
    public PacketSerializer(final Serializer messageSerializer) {
        super(messageSerializer);
    }

    @Override public Packet read(final Reader reader) throws Exception {
        final var requestNumber = reader.readInt();
        return Packet.isEnd(requestNumber) ? Packet.END : new Packet(requestNumber, (Message)serializer.read(reader));
    }

    @Override public void write(final Object value, final Writer writer) throws Exception {
        final var packet = (Packet)value;
        if (packet.isEnd()) {
            writer.writeInt(Packet.END_REQUEST_NUMBER);
        } else {
            writer.writeInt(packet.requestNumber());
            serializer.write(packet.message(), writer);
        }
    }

}

File: kotlin/yass/main/ch/softappeal/yass/transport/PacketSerializer.kt
package ch.softappeal.yass.transport

import ch.softappeal.yass.remote.*
import ch.softappeal.yass.remote.session.*
import ch.softappeal.yass.serialize.*

fun packetSerializer(messageSerializer: Serializer) = object : Serializer {
    override fun read(reader: Reader): Packet {
        val requestNumber = reader.readInt()
        return if (isEndPacket(requestNumber))
            EndPacket
        else
            Packet(requestNumber, messageSerializer.read(reader) as Message)
    }

    override fun write(writer: Writer, value: Any?) {
        val packet = value as Packet
        if (packet.isEnd)
            writer.writeInt(EndRequestNumber)
        else {
            writer.writeInt(packet.requestNumber)
            messageSerializer.write(writer, packet.message)
        }
    }
}
